port/osx/omrintrospect.c

```diff
@@ -300,9 +321,75 @@ getThreadContext(J9ThreadWalkState *state)
 
 	if (0 == ret) {
 		/* Wait for the signal handler to complete. */
-		if (0 == read(pipeFileDescriptor[0], buffer, 1)) {
+		ret = timedWait(timeout(state->deadline1));
+		if (TIMEOUT == ret) {
+			RECORD_ERROR(state, TIMEOUT, 0);
+		}
+	}
+
+	return ret;
+}
+
+/**
+ * Utility function to calculate remaining time before the deadline elapses.
+ *
+ * @param deadline system time by which processing should have completed
+ *
+ * @return number of seconds before timeout.
+ */
+static int64_t
+timeout(int64_t deadline)
+{
+	int64_t secs = 0;
+	struct timespec spec;
+	if (-1 == clock_gettime(CLOCK_REALTIME, &spec)) {
+		return 0;
+	}
+
+	secs = deadline - spec.tv_sec;
+	if (secs > 0) {
+		return secs;
+	}
+
+	return 0;
+}
+
+/**
+ * This function waits for data on a pipe and will block for at most `seconds` seconds.
+ *
+ * @param seconds the maximum number of seconds to block for, must be positive
+ *
+ * @return 0 if success, -1 if error, TIMEOUT if timeout occurred.
+ */
+static int
+timedWait(int64_t seconds)
+{
+	char buffer[1];
+	struct pollfd fds[1];
+	int ret = 0;
+	fds[0].fd = pipeFileDescriptor[0];
+	fds[0].events = POLLIN;
+
+	if (seconds > 0) {
+		ret = poll(fds, 1, (int)seconds * 1000);
```

Comment:
`int64_t` is an inappropriate type for `seconds` if it must nor be more than `INT_MAX / 1000`; suggest `int32_t`.